libraries/chain/controller.cpp

```diff
@@ -2300,21 +2303,53 @@ struct controller_impl {
             }
          }
 
-         snapshot->read_section<value_t>([this,&row_counter]( auto& section ) {
+         snapshot->read_section<value_t>([this,&rows_loaded]( auto& section ) {
             bool more = !section.empty();
             while(more) {
                decltype(utils)::create(db, [this, &section, &more]( auto &row ) {
                   more = section.read_row(row, db);
                });
-               row_counter.progress();
+               rows_loaded.fetch_add(1u, std::memory_order_relaxed);
             }
          });
       });
 
-      read_contract_tables_from_snapshot(snapshot, row_counter);
+      snapshot_load_ctx.post([this,&snapshot,&rows_loaded]() {
+         read_contract_tables_from_snapshot(snapshot, rows_loaded);
+      });
+
+      authorization.read_from_snapshot(snapshot, rows_loaded, snapshot_load_ctx);
+      resource_limits.read_from_snapshot(snapshot, rows_loaded, snapshot_load_ctx);
+
+      constexpr unsigned max_snapshot_load_threads = 4;
+      const unsigned snapshot_load_threads = snapshot->supports_threading() ? max_snapshot_load_threads : 1;
+
+      std::vector<std::promise<void>> thread_promises(snapshot_load_threads);
+      std::list<std::thread> threads;
+      //this scoped_exit can go away with jthread; but still marked experimental in libc++18
+      auto join = fc::make_scoped_exit([&threads]() {
+         for(std::thread& t : threads)
+            t.join();
+      });
+      for(unsigned i = 0; i < snapshot_load_threads; ++i)
+         threads.emplace_back([&snapshot_load_ctx, &prom = thread_promises[i]] {
+            try {
+               snapshot_load_ctx.run();
+               prom.set_value();
+            }
+            catch(...) {
+               snapshot_load_ctx.stop();
+               prom.set_exception(std::current_exception());
+            }
+         });
 
-      authorization.read_from_snapshot(snapshot, row_counter);
-      resource_limits.read_from_snapshot(snapshot, row_counter);
+      for(std::promise<void>& p : thread_promises) {
+         std::future<void> f = p.get_future();
+         while(f.wait_for(std::chrono::seconds(5)) != std::future_status::ready) {
+            ilog("Snapshot initialization ${pct}% complete", ("pct",(unsigned)(((double)rows_loaded/total_snapshot_rows)*100)));
+         }
+         f.get();
+      }
```

Comment:
Would be nice to package the thread pool run to completion into a class in thead_utils.hpp. Either enhance `named_thread_pool` to have this additional capability (provided ctx and set promise when run complete), or add a new class that just does this functionality. Bonus points for setting the thread name if you add a new class.

Reply:
def thinks this should be separate from `named_thread_pool` instead of trying to kludge in two modes of operation. I'll see what I can do